### ISSUE TITLE
AKS 2019-03-21 release

### ISSUE THREAD

**Issue by patrickmslatteryvt** [Azure/AKS#889]
I note that there is a new release of AKS at: https://github.com/Azure/AKS/releases/tag/2019-03-21
But nowhere does it mention what we need to do (if anything) to get this version on our existing AKS clusters.
* Do we do a scale and replace of the nodes?
* Do we just update the the latest K8S release?
* Do we only get this release if we build new clusters?

Any info would be appreciated.
Thanks

@jnoller

**Comment by jnoller** [comment]
Per normal AKS releases, you should only need to upgrade unless otherwise noted.

**Comment by tomasr** [comment]
@jnoller That seems kind of weird.... so you need to upgrade to a different Kubernetes version to get the platform changes? (which would also mean that if you're already on the latest version available you're stuck until a newer Kubernetes release is deployed)

**Comment by patrickmslatteryvt** [comment]
@jnoller So I have a v1.12.6 cluster that is 5 days old and this release was dropped 3 days ago, if I'm understanding you correctly, I would have to rebuild my cluster to get the updates included in this latest AKS release.
Is that correct?

**Comment by jnoller** [comment]
@tomasr Yes, currently that is an existing/known limitation.

@patrickmslatteryvt Not even remotely, you just need to run a normal cluster upgrade.

**Comment by patrickmslatteryvt** [comment]
@jnoller So @tomasr and I asked basically the same question and we got two very different answers.
Would you care to try that answer again?

> you just need to run a normal cluster upgrade.
To what version? **I already have** v1.12.6, I do not see a newer version of K8S available in AKS.

**Comment by jnoller** [comment]
@patrickmslatteryvt I was responding to you individually, to summarize:

1. AKS versions available for users to upgrade to are currently hard-tied to the version of Kubernetes. This is a known issue we are working on - we agree this is confusing and present limitations (as you noted) as in order to apply upgrades there must be a newly available version.

2. For "We have an existing Kubernetes cluster at the latest released kubernetes version", you can 'work around' not having a newer Kubernetes release by applying a same-version upgrade (eg 1.12.6 to 1.12.6), or scale-up down. 

2.1: This operation (same version upgrade) was blocked previously, and was enabled in the March releases. We are still working on breaking up the AKS / kubernetes / aks-engine and other versions to provide discreet semver compatible versioning for the full stack.

3. All other upgrades (not on latest k8s minor release in the version list) get the updates via normal upgrades.

Apologies for the confusion - you're right @patrickmslatteryvt I mixed up the replies / questions.

**Comment by jnoller** [comment]
Correction: And I apologize again for this - there's been a lot of churn here in the code base (to support versioning and other work) as well as work to fix edge cases in the upgrade path for failed states.

Point 2 is incorrect - same version upgrades are currently blocked unless the upgrade failed in a previous attempt. I've edited the comment to remove it.

The reason for this is tied to the component version issue mentioned - eg an in-place upgrade does not force all components to upgrade. We've seen this with Moby/Docker runtimes in the past - this is why it's currently blocked as in the correct solution, if a user requests an upgrade from X to X (where X is the major.minor.patch) should refresh/update all component versions.

Currently same-version upgrades are supported, but **only** in the case where a user requested an upgrade but it failed for a secondary reason (eg. lack of quota, subnet, etc).

So to recap, yes - @patrickmslatteryvt if you are on Kubernetes 1.12.6 in order to get the service updates you have to perform a scale-out.

**Comment by jnoller** [comment]
Filed https://github.com/Azure/AKS/issues/893 to support same-version 'upgrades' / refresh

**Comment by jnoller** [comment]
And here is the docs bug (on this repo/assigned to me) https://github.com/Azure/AKS/issues/894

**Comment by patrickmslatteryvt** [comment]
Thank you for the detailed explanation 👍
